File: rust/uniswap/src/event_handlers.rs
use chaindexing::states::{ContractState, Filters, MultiChainState, Updates};
use chaindexing::{EventContext, EventHandler};

use crate::states::{Pool, TokenSwapVolume};

pub struct PoolCreatedEventHandler;

#[chaindexing::augmenting_std::async_trait]
impl EventHandler for PoolCreatedEventHandler {
    fn abi(&self) -> &'static str {
        "PoolCreated(address indexed token0, address indexed token1, uint24 indexed fee, int24 tickSpacing, address pool)"
    }

    async fn handle_event<'a, 'b>(&self, context: EventContext<'a, 'b>) {
        let event_params = context.get_event_params();

        // Extract each parameter as exactly specified in the ABI:
        let token0_address = event_params.get_address_string("token0");
        let token1_address = event_params.get_address_string("token1");
        let pool_contract_address = event_params.get_address_string("pool");
        let fee = event_params.get_u32("fee");
        let tick_spacing = event_params.get_i32("tickSpacing");

        Pool {
            token0_address,
            token1_address,
            pool_contract_address: pool_contract_address.clone(),
            fee,
            tick_spacing,
        }
        .create(&context)
        .await;

        // Include new UniswapV3Pool contract:{pool_contract_address} for indexing...
        chaindexing::include_contract(&context, "UniswapV3Pool", &pool_contract_address).await;
    }
}

pub struct SwapEventHandler;

#[chaindexing::augmenting_std::async_trait]
impl EventHandler for SwapEventHandler {
    fn abi(&self) -> &'static str {
        "Swap(address indexed sender, address indexed recipient, int256 amount0, int256 amount1, uint160 sqrtPriceX96, uint128 liquidity, int24 tick)"
    }

    async fn handle_event<'a, 'b>(&self, context: EventContext<'a, 'b>) {
        let event_params = context.get_event_params();
        let event = &context.event;

        let pool = Pool::read_one(
            &Filters::new("pool_contract_address", &event.contract_address).within_multi_chain(),
            &context,
        )
        .await
        .unwrap();

        // Extract each parameter as exactly specified in the ABI:
        let amount0 = event_params.get_int_ether("amount0");
        let amount1 = event_params.get_int_ether("amount1");

        let token0 = pool.token0_address;
        let token1 = pool.token1_address;

        let last_updated_at = event.get_block_timestamp();

        create_or_update_token_swap_volume(token0, amount0, last_updated_at, &context).await;
        create_or_update_token_swap_volume(token1, amount1, last_updated_at, &context).await;
    }
}

async fn create_or_update_token_swap_volume<'a, 'b>(
    token_address: String,
    amount_ether: f64,
    last_updated_at: u64,
    context: &EventContext<'a, 'b>,
) {
    let amount_ether = amount_ether.abs(); // Volume calculation is additive

    if let Some(prev_token_swap_volume) = TokenSwapVolume::read_one(
        &Filters::new("token_address", &token_address).within_multi_chain(),
        context,
    )
    .await
    {
        let new_amount_ether =
            prev_token_swap_volume.amount_ether.parse::<f64>().unwrap() + amount_ether;
        let updates =
            Updates::new("amount_ether", new_amount_ether).add("last_updated_at", last_updated_at);

        prev_token_swap_volume.update(&updates, context).await;
    } else {
        TokenSwapVolume {
            token_address,
            amount_ether: amount_ether.to_string(),
            last_updated_at,
        }
        .create(context)
        .await;
    }
}

File: typescript/uniswap/src/event-handlers.ts
import {
  PureHandler,
  PureHandlerContext,
} from '../../../../chaindexing-ts/chaindexing-core/src';
import { Pool, TokenSwapVolume } from './states';

export class PoolCreatedEventHandler implements PureHandler {
  abi(): string {
    return 'event PoolCreated(address indexed token0, address indexed token1, uint24 indexed fee, int24 tickSpacing, address pool)';
  }

  async handleEvent(context: PureHandlerContext): Promise<void> {
    const eventParams = context.getEventParams();

    // Extract each parameter as exactly specified in the ABI
    const token0Address = eventParams.getAddressString('token0');
    const token1Address = eventParams.getAddressString('token1');
    const poolContractAddress = eventParams.getAddressString('pool');
    const fee = eventParams.getU32('fee');
    const tickSpacing = eventParams.getU32('tickSpacing'); // Note: treating as unsigned for now

    console.log(`Creating pool: ${token0Address}/${token1Address} (fee: ${fee}, pool: ${poolContractAddress})`);

    // Create new pool state
    const pool = new Pool(
      token0Address,
      token1Address,
      fee,
      tickSpacing,
      poolContractAddress
    );

    await pool.create(context);

    console.log(`✅ Created pool ${poolContractAddress} for tokens ${token0Address}/${token1Address}`);

    // TODO: In the full implementation, we would also include the new pool contract for indexing
    // This would be equivalent to: chaindexing::include_contract(&context, "UniswapV3Pool", &pool_contract_address)
    // For now, we'll log that this would happen
    console.log(`📝 Would include contract ${poolContractAddress} for future indexing`);
  }
}

export class SwapEventHandler implements PureHandler {
  abi(): string {
    return 'event Swap(address indexed sender, address indexed recipient, int256 amount0, int256 amount1, uint160 sqrtPriceX96, uint128 liquidity, int24 tick)';
  }

  async handleEvent(context: PureHandlerContext): Promise<void> {
    const eventParams = context.getEventParams();
    const event = context.event;

    console.log(`Processing swap in pool: ${event.contractAddress}`);

    // TODO: In the full implementation, we would read the pool data
    // For now, we'll simulate the pool data (this would normally come from the database)
    const simulatedPool = {
      token0Address: '0xA0b86a33E6441e05e60BDA6C00C0b0D07c7e7b5C', // Mock token0
      token1Address: '0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2', // Mock token1 (WETH)
    };

    // Extract swap amounts (using BigInt for large numbers, then converting to number)
    const amount0 = Number(eventParams.getBigInt('amount0')) / 1e18; // Convert from wei to ether
    const amount1 = Number(eventParams.getBigInt('amount1')) / 1e18; // Convert from wei to ether

    const lastUpdatedAt = event.blockTimestamp;

    console.log(`Swap amounts: ${amount0} (token0), ${amount1} (token1)`);

    // Update volume for both tokens
    await this.createOrUpdateTokenSwapVolume(
      simulatedPool.token0Address,
      Math.abs(amount0),
      lastUpdatedAt,
      context
    );

    await this.createOrUpdateTokenSwapVolume(
      simulatedPool.token1Address,
      Math.abs(amount1),
      lastUpdatedAt,
      context
    );

    console.log(`✅ Updated swap volumes for tokens ${simulatedPool.token0Address} and ${simulatedPool.token1Address}`);
  }

  private async createOrUpdateTokenSwapVolume(
    tokenAddress: string,
    amountEther: number,
    lastUpdatedAt: number,
    context: PureHandlerContext
  ): Promise<void> {
    // TODO: In the full implementation, we would read existing volume from database
    // For now, we'll always create new records (this would normally update existing)
    
    console.log(`Creating/updating volume for token ${tokenAddress}: ${amountEther} ETH`);

    const tokenSwapVolume = new TokenSwapVolume(
      tokenAddress,
      amountEther.toString(),
      lastUpdatedAt
    );

    await tokenSwapVolume.create(context);

    console.log(`📊 Updated volume for ${tokenAddress}: ${amountEther} ETH`);
  }
} 